### PR TITLE
Initial Implementation of Ext4 Module

### PR DESCRIPTION
**Pull request by movitto**
Based on #9 this enhances virtfs-ext3 to be able to handle ext4 logic and adds virtfs-ext4 to implement the extended format.

Contains small fixes & style enhancements to unify / bring everything together.

### CODE REVIEW

**Comment by Fryguy** [comment]
@roliveri Should this be a separate gem?  Or do you want to commit here first and then transfer later?

File: lib/virtfs/core_ext.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class String
+  def UnicodeToUtf8
```
Comment (Fryguy) [inline lib/virtfs/core_ext.rb]: @roliveri Moving forward, is there a way we can do this without monkey patching the string class?  I'm also concerned that if it's already patched elsewhere, then this code ignores that. Perhaps it should be wrapped in an `unless defined?(UnicodeToUtf8)`

**Comment by Fryguy** [comment]
@movitto Is this a straight copy from ManageIQ, with no modifications?  If there are modifications, what did you change?  Also, if it came from ManageIQ, the license needs to be Apache 2.0.

Also also, the tests are failing both for this PR and #9 .  If you need help getting them green, let me know.

File: virtfs-ext3/virtfs-ext3.gemspec [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  else
+    raise "RubyGems 2.0 or newer is required to protect against public gem pushes."
+  end
+
+  spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
+  spec.bindir        = "exe"
+  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
+  spec.require_paths = ["lib"]
+
+  spec.add_runtime_dependency "virtfs"
```
Comment (Fryguy) [inline virtfs-ext3/virtfs-ext3.gemspec:31]: If this gem depends on the ext3 gem, then there should probably be a depedency line here.

**Comment by roliveri** [comment]
@Fryguy It should be separate gem. Can you create a repo for it? Then Mo can push his branch up there.

**Comment by movitto** [comment]
@Fryguy @roliveri yes we need separate gems for virtfs-ext3, virtfs-ext4, and virtfs-ntfs. Also others could be created in advance for: reiserfs, xfs, aufs, zfs, refs, apfs, webdav, btfs.

License has been changed. Tests have been fixed (missing version.rb files). 

This was mostly a straightforward extraction from manageiq though included style updates to conform to our latest standards (underscoring camelcased methods, 'rubyizing' method names, etc). 

Let me know if there is anything else

**Comment by movitto** [comment]
Closing this out as this was now sent to virtfs-ext4

**Pull request closed without merging**